Created `Current-guide.md` (#86)
Babylonpartners/ios-playbook: Cookbook/Current.md

# How to use `Current` in Babylon iOS Project

### Where is that? 
- `Current` is the global instance of the `World` type which is located in the `BabylonDependencies.framework`. 

> It's purpose is to provide convenient access to the instances that should be shared between different parts of the app (cross-cutting concerns). For more information please read the [How Control the World](/Cookbook/Proposals/ControlTheWorld.md) document.

Reviewer: ```suggestion
Its purpose is to provide convenient access to the instances that should be shared between different parts of the app (cross-cutting concerns). For more information please read the [How Control the World](/Cookbook/Proposals/ControlTheWorld.md) document.
```

Could we add a few examples? Just to get an idea, because this is very abstract talk and I don't think I really understand it 😅

Author: the link explains it with loads of examples. it'll be redundant.

## `Do`s and `Don't`s

> #### do

Reviewer: ```suggestion
#### Dos
```

- access `Current` **only** from `Builder`s layer;

> - pass the required dependencies down the chain;

Reviewer: What does this mean? Using current to access some stuff that is then injected to a VM?

Author: pass the dependencies to VM and any other layer that needs any dependency. 
Isn't that clear?

> - contribute to it;

Reviewer: How? What can we add to it and what should we not?

Author: I think it doesn't make sense.. once all `appDependencies` are moved into Current, no need to contribute to it. As Giorgos said, it's a matter to follow the pattern os other properties: with default value or set asynchronously.

I changed to the mutation when it's allowed. 😉

> #### don't

Reviewer: ```suggestion
#### Don't
```

> - access `Current` from any other layer other than `Builder`. I.E all the other layers (ViewModel, Flowcontroller, ViewController, Model, BusinessController) continue to receive its dependencies throught injection, preferably via `init`;

Reviewer: ```suggestion
- access `Current` from any other layer other than `Builder`. All the other layers (`ViewModel`, `FlowController`, `ViewController`, `Model`, `BusinessController`) continue to receive their dependencies through injection when created by the builder, preferably at initialisation time;
```

> - inject `current` as a dependency;

Reviewer: ```suggestion
- inject `Current` as a dependency;
```

Author: it's intentional to be `current` because I already saw it appearing in the codebase as lowerCase.

> - try to mutate `Current`. In `Release` configuration it's a constant `let`;

Reviewer: Could this be further explained? For example I didn't really follow the release issue on slack, so I have no idea why the Release config has to be a constant property.

Author: it's basically this bit: 
```swift
#if DEBUG
public var Current: World = .production
#else
public let Current: World = .production
#endif
```

We can mutate for Debug and specially for Tests. But production not at all. 
But it mutates via `setPropertyName` which changes the value of `MutableProperties` which maintains the immutability of `Current`. 
Semantics ✨ 

I'll think how that can be explained in there.
👍

Author: just to highlight

`public var`
`public let` 

I took time to realise the difference of these 2 lines. lol

Author: added the snippet

# How to use `Current` in Babylon iOS Project

> 

Reviewer: I would explain what Current is and what its purpose is first.

Author: https://github.com/Babylonpartners/ios-playbook/blob/master/Cookbook/Proposals/ControlTheWorld.md

Author: also this is a TBD.

Reviewer: Yes, but that is a proposal. This is the documentation for it -- I think it should serve as an independent and self-sufficient document explaining what Current is, what it does, how to use it, future directions, etc.

Reviewer: > also this is a TBD contents.
Then maybe we should have this as WIP?

Author: I agree. But it's the beginning. It'll grow.

Reviewer: I feel like at the current state, and as it is public, this document is a bit too barebones to be merged.

Author: I see no problem: 
- there's no link to this document;
- it's a documentation and there's nothing wrong on the content and publishing an in progress document;
- life is asynchronous 😛

Author: @diegopetrucci We need to speed up documentations.

Reviewer: But you still have to go through PR approval whenever you want to make a change, so why not beefing it up a bit more for the first draft?

Author: the first idea was to create *just* the title.. and I did a lot more.

Author: > I would explain what Current is and what its purpose is first.

there's a reference to the proposal at the bottom.

### Where is that? 

> - `Current` is the global implementation of the `World` file which is located in the `BabylonDependencies.framework`.

Reviewer 2: ```suggestion
- `Current` is the global instance of the `World` type which is located in the `BabylonDependencies.framework`. It's purpose is to provide convenient access to the instances that should be shared between different parts of the app (cross-cutting concerns)
```

> ### Properties with default values

Reviewer 3: What is the value of documenting these here?

Author: you will see when you need to create one here.

Author: it's a guide

TBD

> ### Properties without default values..

Reviewer 3: What is the value of documenting these here?

Author: ditto

..which depends on asyncronous events to have its value

TBD

> ### Runtime dynamic changes as `DesignLibrary`

Reviewer 3: What is the value of documenting these here?
If the variable is something Reactive like `Signal`/`Property`/etc you get a strong hint 
that it will change over time.

Author: > strong hint 

is not documentation.

If you feel that it's overlapping too much, I can remove it from here. 
But I feel that it's not that clear how to use stuff (Current, DesignLibrary, NVL, components). Imagine for a new joiner. 

I thought you'd feel happy with it. 

Again, don't you see value on this kind of documentation? @gtsiap

Reviewer 2: I think it will make sense at least to mention that these properties should be observed rather than be accessed via `.value`. But for that may the next section will be already a good fit.
